felix/bpf/conntrack/v5/map6.go

```diff
@@ -0,0 +1,442 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v4
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/felix/bpf/maps"
+)
+
+//	struct calico_ct_key {
+//	  uint32_t protocol;
+//	  __be32 addr_a, addr_b; // NBO
+//	  uint16_t port_a, port_b; // HBO
+//	};
+const KeyV6Size = 40
+const ValueV6Size = 128
+
+type KeyV6 [KeyV6Size]byte
+
+func (k KeyV6) AsBytes() []byte {
+	return k[:]
+}
+
+func (k KeyV6) Proto() uint8 {
+	return uint8(binary.LittleEndian.Uint32(k[:4]))
+}
+
+func (k KeyV6) AddrA() net.IP {
+	return k[4:20]
+}
+
+func (k KeyV6) PortA() uint16 {
+	return binary.LittleEndian.Uint16(k[36:38])
+}
+
+func (k KeyV6) AddrB() net.IP {
+	return k[20:36]
+}
+
+func (k KeyV6) PortB() uint16 {
+	return binary.LittleEndian.Uint16(k[38:40])
+}
+
+func (k KeyV6) String() string {
+	return fmt.Sprintf("ConntrackKey{proto=%v %v:%v <-> %v:%v}",
+		k.Proto(), k.AddrA(), k.PortA(), k.AddrB(), k.PortB())
+}
+
+func (k KeyV6) Upgrade() maps.Upgradable {
+	panic("conntrack map key already at its latest version")
+}
+
+func NewKeyV6(proto uint8, ipA net.IP, portA uint16, ipB net.IP, portB uint16) KeyV6 {
+	var k KeyV6
+	binary.LittleEndian.PutUint32(k[:4], uint32(proto))
+	copy(k[4:20], ipA.To16())
+	copy(k[20:36], ipB.To16())
+	binary.LittleEndian.PutUint16(k[36:38], portA)
+	binary.LittleEndian.PutUint16(k[38:40], portB)
+	return k
+}
+
+// struct calico_ct_value {
+//  __u64 rst_seen;
+//  __u64 last_seen; // 8
+//  __u8 type;     // 16
+//  __u8 flags;     // 17
+//
+//  // Important to use explicit padding, otherwise the compiler can decide
+//  // not to zero the padding bytes, which upsets the verifier.  Worse than
+//  // that, debug logging often prevents such optimisation resulting in
+//  // failures when debug logging is compiled out only :-).
+//  __u8 pad0[5];
+//  __u8 flags2;
+//  union {
+//    // CALI_CT_TYPE_NORMAL and CALI_CT_TYPE_NAT_REV.
+//    struct {
+//      struct calico_ct_leg a_to_b; // 24
+//      struct calico_ct_leg b_to_a; // 36
+//
+//      // CALI_CT_TYPE_NAT_REV only.
+//      __u32 orig_dst;                    // 48
+//      __u16 orig_port;                   // 52
+//      __u8 pad1[2];                      // 54
+//      __u32 tun_ip;                      // 56
+//      __u32 pad3;                        // 60
+//    };
+//
+//    // CALI_CT_TYPE_NAT_FWD; key for the CALI_CT_TYPE_NAT_REV entry.
+//    struct {
+//      struct calico_ct_key nat_rev_key;  // 24
+//      __u8 pad2[8];
+//    };
+//  };
+// };
+
+const (
+	VoRSTSeenV6   int = 0
+	VoLastSeenV6  int = 8
+	VoTypeV6      int = 16
+	VoFlagsV6     int = 17
```

Comment:
Please change these constants as per new value struct.

Reply:
Can you check them now? I think I've updated them right, but could do with another pair of eyes